File: backend/app/schemas/analytics.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class AnalyticsParams(BaseModel):
    """Filters for the dashboard.

    Deliberately a subset of the directory's: there is no free-text search,
    because a dashboard answers "how do we pay this group" rather than "find this
    person", and no status filter, because the totals are always about people
    currently being paid.
    """

    country_id: list[int] = Field(default_factory=list)
    department_id: list[int] = Field(default_factory=list)
    job_level_id: list[int] = Field(default_factory=list)


class TotalsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    headcount: int
    total_spend_usd: Decimal
    # Null rather than zero when no one matches: an average of nothing is not 0.
    average_salary_usd: Decimal | None
    median_salary_usd: Decimal | None
    leaving_soon: int


class BreakdownOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    headcount: int
    total_spend_usd: Decimal
    average_salary_usd: Decimal


class AnalyticsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    totals: TotalsOut
    by_country: list[BreakdownOut]
    by_department: list[BreakdownOut]

File: frontend/src/api/analytics.ts
import { api, buildUrl } from "./client";

/** Money arrives as decimal strings for the same reason salaries do: parsing to
 *  a JS float would reintroduce the rounding the backend works to avoid. */
export type Totals = {
  headcount: number;
  total_spend_usd: string;
  /** Null when no one matches the filters. An average of nothing is not zero. */
  average_salary_usd: string | null;
  median_salary_usd: string | null;
  /** Active employees with a departure already scheduled. */
  leaving_soon: number;
};

export type Breakdown = {
  id: number;
  name: string;
  headcount: number;
  total_spend_usd: string;
  average_salary_usd: string;
};

export type Analytics = {
  totals: Totals;
  by_country: Breakdown[];
  by_department: Breakdown[];
};

/** A subset of the directory's filters: no search, and no status, since the
 *  dashboard always reports on people currently being paid. */
export type AnalyticsFilters = {
  country_id?: number[];
  department_id?: number[];
  job_level_id?: number[];
};

export function fetchAnalytics(filters: AnalyticsFilters): Promise<Analytics> {
  return api.get<Analytics>(buildUrl("/analytics", { ...filters }));
}
